tests/unit/conftest.py

```diff
@@ -72,3 +72,18 @@ def bugtool(imported_bugtool):
     imported_bugtool.data = {}
     sys.argv = ["xen-bugtool", "--unlimited"]
     return imported_bugtool
+
+
+@pytest.fixture(scope="function")
+def tmpdir(tmp_path):
+    """Run test in a temporary directory and check leftovers after the test."""
+    temp_dir = tmp_path.as_posix()
+
+    curdir = os.getcwd()
+    os.chdir(temp_dir)
+    yield temp_dir
```

Comment:
This is clever (I hope I understand it correctly)

Reply:
Yes, it'm 💯 % sure you do! 🥇 :)

It's like a context manager or like a Generator, but as a `pytest` fixture which the tests request using a `@pytest.marker.usefixture("tmpdir")` if they don't need the path name of the directory but just are happy with using it, and they get check added by the fixture which checks that the testcase execution does not leave files around after exiting. (and they can also use the temp_dir variable if they need, but they don't need it in this case.)

Reply:
Does this depend on the caller using `next` or `yield` itself or is this implicit and control returns at some other defined point?

Reply:
It is a fixture for pytest with scope "function".

Because the scope is function for this fixture, the scope of the yield is one complet test case function.

Pytest fixtures are used as a layer between pytest itself and the test case function:

I updated it now to not override the built-in tmpdir function, now the simples form is in_tmpdir().

In the simplest way, it would look like this:

```py
| pytest finds all test case functions, looks up the fixtures they want.
\- tmpdir (built into pytest) provides the path to temporary directory
 \- in_tmpdir changes to the tmpdir, yields to the requester and on exit, changes back
  \- test_function runs in the tmpdir
  /- test_function returns
 /- in_tmpdir changes back to the regular working directory
/- tmpdir (built into pytest) cleans up the tmpdir
|- pytest itself gets control back, depending on its config will show results from the case
\-tmpdir creats another tmpdir
 \- another fixture like doing some reset on the tested module for the next test runs
  \- a testcase that expects to be in the srcdir runs
  /- the testcase returns
 /- the other fixture does its cleanup if needed (if it used yield)
| pytest runs again
```
... and so on.

Updates pushed:
- Improved to use the built-in `tmpdir` fixture, this fixture now has the name `in_tmpdir`.
- a new sub-fixture bugtool_log() is now added.
  - it is ensuring that the bugtool log that allows for logging of errors is checked
  - the job of checking of the code under test left files by accident is moved to it
- add a new fixture that detects even when bugtool tries to use the current directory to create temporary files. This is how the implementation of `log()` works. Aims to not let such canges go unnoticed by tests in the future.
- And the tests are updated to use the new fixure. This removes the need for boilerplate decorators just to launch teset cases which want it using a decorator or another fixture. They are now run inside their own private tmpdir solely by importing the bugtool fixture that thay need to run their testcode, very nice! :)